javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll

```diff
@@ -305,18 +305,64 @@ module ReflectedXss {
    * a content type that does not (case-insensitively) contain the string "html". This
    * is to prevent us from flagging plain-text or JSON responses as vulnerable.
    */
-  private class HttpResponseSink extends Sink, DataFlow::ValueNode {
+  class HttpResponseSink extends Sink, DataFlow::ValueNode {
     override HTTP::ResponseSendArgument astNode;
 
-    HttpResponseSink() { not nonHtmlContentType(astNode.getRouteHandler()) }
+    HttpResponseSink() { not exists(getAnonHtmlHeaderDefinition(astNode)) }
+  }
+
+  /**
+   * Gets a HeaderDefinition that defines a non-html content-type for `send`.
+   */
+  HTTP::HeaderDefinition getAnonHtmlHeaderDefinition(HTTP::ResponseSendArgument send) {
+    exists(HTTP::RouteHandler h |
+      send.getRouteHandler() = h and
+      result = nonHtmlContentTypeHeader(h)
+    |
+      // not the case that the control just exists without potentially going to the worksFor.
+      not isIrrelevantFor(result, send)
+    )
   }
 
   /**
    * Holds if `h` may send a response with a content type other than HTML.
    */
-  private predicate nonHtmlContentType(HTTP::RouteHandler h) {
-    exists(HTTP::HeaderDefinition hd | hd = h.getAResponseHeader("content-type") |
-      not exists(string tp | hd.defines("content-type", tp) | tp.regexpMatch("(?i).*html.*"))
+  HTTP::HeaderDefinition nonHtmlContentTypeHeader(HTTP::RouteHandler h) {
+    result = h.getAResponseHeader("content-type") and
+    not exists(string tp | result.defines("content-type", tp) | tp.regexpMatch("(?i).*html.*"))
+  }
+
+  /**
+   * Holds if a header set in `header` is unlikely to affect a resonse send in `sender`.
+   */
+  predicate isIrrelevantFor(HTTP::HeaderDefinition header, HTTP::ResponseSendArgument sender) {
```

Comment:
I'm not comfortable with this predicate.

In terms of  performance, it appears to be a cartesian product of all header definitions and sender arguments with some tuples subtracted. Indeed, the second disjunct below doesn't bind `sender`.

I'm also a little worried about the second disjunct being too dependent on everything being in a final basic block containing no resolved calls.

Have you tried using dominance for the local control-flow checks? I think we could define it all in terms of dominance relations, like:
- A send is affected by its dominating header definition.
- A dominating header definition is "local" if it dominates the response `end` call.
- A send without a dominating header definition is assumed to be affected by any non-local header definition on the same response object (i.e. a flow-insensitive heuristic for the non-local case).

Lastly, content-type tracking seems like an important enough concept to justify support in the `HTTP` library, so if you feel like building a more detailed analysis (akin to the middleware-tracking in Express) I'd be in favor of that.

Reply:
> Have you tried using dominance for the local control-flow checks? I think we could define it all in terms of dominance relations, like:
> 
> * A send is affected by its dominating header definition.
> * A dominating header definition is "local" if it dominates the response `end` call.
> * A send without a dominating header definition is assumed to be affected by any non-local header definition on the same response object (i.e. a flow-insensitive heuristic for the non-local case).

I tried out an implementation based on your idea, and it works nicely :+1: 

> Lastly, content-type tracking seems like an important enough concept to justify support in the `HTTP` library, so if you feel like building a more detailed analysis (akin to the middleware-tracking in Express) I'd be in favor of that.

I don't see the need for such a detailed analysis yet. `Content-type` tracking is currently only used for this `XSS` sink, so I want to wait with such an analysis until we have another use for `Content-type`.